### PR TITLE
LSM: Improve fuzzer for KWayMergeIterator

### PR DESCRIPTION
**Pull request by toziegler**
This PR improves the fuzzer in two ways:

1. It now exercises all cases that can return `Drained`.
2. It now tests the `reset` method, ensuring that de-duplication still works correctly after a reset.

Both are important, since these behaviors are directly relied upon in `scan_tree.zig` and `scan_merge.zig`.

Previously, these cases were only tested indirectly through other fuzzers. By documenting them in the data structure itself, we ensure that unit tests are in place, making future implementation changes faster.

### CODE REVIEW

File: src/lsm/k_way_merge.zig [hunk 1]
```diff
@@ -95,7 +95,6 @@ pub fn KWayMergeIteratorType(
 
             // We must loop on stream_index but assign at it.k, as k may be less than stream_index
             // when there are empty streams.
-            // TODO Do we have test coverage for this edge case?
```
Comment (batiati) [inline src/lsm/k_way_merge.zig:93]: :heart:

**Comment by toziegler** [comment]
Rebased due to #3217

**Review by batiati — APPROVED** [review]
